aspnetcore/blazor/webassembly-performance-best-practices.md

```diff
@@ -11,116 +11,537 @@ uid: blazor/webassembly-performance-best-practices
 
-          var value = jsInProcess.InvokeUnmarshalled<string>("jsInteropCall");
-      }
-  }
-  ```
+For more information, see <xref:blazor/components#attribute-splatting-and-arbitrary-parameters>.
+
+#### Implement SetParametersAsync manually
+
+One of the main aspects of the per-component rendering overhead is writing incoming parameter values to the `[Parameter]` properties. The renderer has to use reflection to do this. Even though this is somewhat optimized, the absence of JIT support on the WebAssembly runtime imposes limits.
+
+In some extreme cases, you may wish to avoid the reflection and implement your own parameter setting logic manually. This may be applicable when:
+
+ * You have a component that renders extremely often (e.g., there are hundreds or thousands of copies of it in the UI)
+ * It accepts many parameters
+ * You find that the overhead of receiving parameters has an observable impact on UI responsiveness
+
+In these cases you can override the component's virtual <xref:Microsoft.AspNetCore.Components.ComponentBase.SetParametersAsync%2A> method and implement your own component-specific logic. The following example is optimized aggressively to avoid any dictionary lookups or long sequences of string comparisons:
+
+```razor
+@code {
+    [Parameter] public int MessageId { get; set; }
+    [Parameter] public string Text { get; set; }
+    [Parameter] public EventCallback<string> TextChanged { get; set; }
+    [Parameter] public Theme CurrentTheme { get; set; }
+
+    public override Task SetParametersAsync(ParameterView parameters)
+    {
+        foreach (var parameter in parameters)
+        {
+            WriteSingleParameter(parameter.Name, parameter.Value);
+        }
+
+        // Run the normal lifecycle methods, but without assigning parameters again
+        return base.SetParametersAsync(ParameterView.Empty);
+    }
+
+    private void WriteSingleParameter(string name, object value)
+    {
+        switch (name[0])
+        {
+            case 'C':
+                if (name == nameof(CurrentTheme))
+                {
+                    CurrentTheme = (Theme)value;
+                    return;
+                }
+                break;
+            case 'M':
+                if (name == nameof(MessageId))
+                {
+                    MessageId = (int)value;
+                    return;
+                }
+                break;
+            case 'T':
+                if (name == nameof(Text))
+                {
+                    Text = (string)value;
+                    return;
+                }
+                else if (name == nameof(TextChanged))
+                {
+                    TextChanged = (EventCallback<string>)value;
+                    return;
+                }
+                break;
+        }
```

Comment:
I'm in two minds about this code snippet. It could be reduced to something simpler like:

```cs
switch (name)
{
    case nameof(CurrentTheme):
        CurrentTheme = (Theme)value;
        break;
    case nameof(MessageId):
        MessageId = (string)value;
        break;
... etc ...
    default:
        throw exception;
}
```

... and the performance would be basically just as good in this case. It's only in more complicated cases in general where it's worth special logic to avoid sequential string comparisons (e.g., you have parameters called `MyParam1`,  `MyParam2`,  `MyParam3` etc, where the string comparison would end up walking through a lot of characters before it could detect non-matches).

I guess I'm talking myself into simplifying this, even at the cost that some developers will produce suboptimal results with it.

@pranavkm Do you have an opinion?

Reply:
It might be worthwhile to start with the simplified example you have in the comment here. Maybe toss in a comment that says this could be further optimized by using a multi-level or hierarchical comparison to minimize string comparison costs. It would be good not to have developers write incredibly messy code when in the vast majority of cases, this is already significantly better than our dictionary implementation.

Related thought experiment: would we consider code-genning this for the user? Maybe an opt-in if we want to evolve the default implementation?

Reply:
OK I'll simplify it.

> Related thought experiment: would we consider code-genning this for the user? Maybe an opt-in if we want to evolve the default implementation?

Yes, definitely. We thought about that in .NET 5 (or at least Javier and I did) but it was quite late in the cycle and we didn't think it fitted into the remaining capacity. As a .NET 6 feature it would definitely make sense.

Reply:
Updated